pkg/config/config.go

```diff
@@ -799,35 +828,67 @@ func ConvertMapToCty(iMap map[string]interface{}) (map[string]cty.Value, error)
 func ResolveVariables(
 	ctyMap map[string]cty.Value,
 	origin map[string]cty.Value,
+	allowedUnknownNames []string,
 ) error {
 	evalCtx := &hcl.EvalContext{
 		Variables: map[string]cty.Value{"var": cty.ObjectVal(origin)},
 	}
+	unknownMap := map[string]bool{}
+	for _, n := range allowedUnknownNames {
+		unknownMap[n] = true
+	}
 	for key, val := range ctyMap {
-		if val.Type() == cty.String {
-			var valString string
-			if err := gocty.FromCtyValue(val, &valString); err != nil {
-				return err
-			}
-			ctx, varName, found := IdentifyLiteralVariable(valString)
-			// only attempt resolution on global literal variables
-			// leave all other strings alone (including non-global)
-			if found && ctx == "var" {
-				varTraversal := hcl.Traversal{
-					hcl.TraverseRoot{Name: ctx},
-					hcl.TraverseAttr{Name: varName},
-				}
-				newVal, diags := varTraversal.TraverseAbs(evalCtx)
-				if diags.HasErrors() {
-					return diags
-				}
-				ctyMap[key] = newVal
-			}
+		newVal, err := cty.Transform(val, func(p cty.Path, v cty.Value) (cty.Value, error) {
+			return resolveValue(v, evalCtx, unknownMap)
+		})
+
+		var re *ResolutionError
+		if errors.As(err, &re) && re.allowed {
```

Comment:
I think it's overly complex way to do it.
Could we instead do something like this without leaking to `ResolveVariables`?:
```go
// packerwtiter
dependOnIgc := findSettingsDependingOnIgc(mod.Settings.Items()) // set of setting names
defaults := removeKeys(mod.Settings.Items(), dependOnIgc) 
defaults = ResolveVariables(defaults, dc.Config.Vars.Items())
writeautovars(defaults)
printPackerInstructions(modPath, mod.ID,  len(dependOnIgc) > 0)
```

Reply:
We will implement this alongside other planned changes in very near-term.